### ISSUE TITLE
[Feature Request] Allowing custom callbacks to scipy optimisation loop

### ISSUE THREAD

**Issue by huylenguyen** [pytorch/botorch#1634]
# 🚀 Feature Request

Currently, `fit_gpytorch_scipy()` has a hardcoded callback function `store_iteration()` which is passed to `scipy.optimize.minimize()` based on the argument `track_iterations`. It would be helpful to expose the `callback` argument of `minimize()` to allow custom user-defined callback functions.

## Motivation

Currently, I am trying to implement a hyperparameter optimization loop for Gaussian Processes to automatically select kernels. I want to do this in [Ray](https://docs.ray.io/en/latest/tune/index.html), as poor kernels can usually be identified early on in model fitting and the corresponding trials trimmed.

This requires `scipy.optimize.minimize()` to accept a custom callback function to report the objective function metric to `ray` at each iteration.

## Pitch

Replace the parameter `track_iterations` in `fit_gpytorch_scipy` with a `callback` parameter which can be `store_iterations()`, `None`, or a user-defined callback. 

If there is interest in developing this as a feature, I would be happy to open a pull request.

**Comment by j-wilson** [comment]
Hi @huylenguyen. `fit_gpytorch_scipy` is marked for deprecation. The most recent code uses `fit_gpytorch_mll_scipy`, which internally calls `botorch.optim.core.scipy_minimize`. Both of these methods expose a callback argument `callback: Callable[[Dict[str, Tensor], OptimizationResult], None]`. Inside of `scipy_minimize`, [this callback get wrapped](https://github.com/pytorch/botorch/blob/3596b125ffad5482eb113ef55f32375ec66ff0ec/botorch/optim/core.py#L100) to take an `ndarray` and passed to `scipy.optimize.minimize`.  

Does this work for your purposes?

**Comment by huylenguyen** [comment]
> Hi @huylenguyen. `fit_gpytorch_scipy` is marked for deprecation. The most recent code uses `fit_gpytorch_mll_scipy`

Hi @j-wilson, thank you very much for bringing this to my attention. I have updated to the latest release version and I see now that the optimizer for `fit_gpytorch_mll` has changed to this new function. This will fit my needs nicely, thank you!
